src/main/java/co/cask/gcp/bigquery/BigQuerySink.java

```diff
@@ -442,4 +451,104 @@ private void decodeSimpleTypes(JsonObject json, String name, StructuredRecord in
         break;
     }
   }
+
+  /**
+   * Validates output schema against bigquery table schema. It throws {@link IllegalArgumentException}
+   * if the output schema has more fields than bigquery table or output schema field types does not match bigquery
+   * column types.
+   */
+  private void validateSchema(BigQuerySinkConfig config) throws IOException {
```

Comment:
We generally put validation of schema and other config properties in the config class itself

Reply:
I have kept it in the same class because it uses BigQuery client to verify the schema. In future if client side requires some changes, it is there in one place. Also in hydrator-plugins we validate schema in the same class as plugin in some plugins: https://github.com/caskdata/hydrator-plugins/blob/develop/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java#L230

Reply:
When this class calls validate it can query and pass the schema of actual table as para. The config validation logic should reside in the config IMO. In Hydrator where it is done outside of the config is not a good practice and should not be used as an example for new code.

Reply:
In general for new code we've been putting a validate() method in the config that can be used to do any validation. There are a few reasons it seems to be easier in the config:

Validation logic can be easily unit tested by simply creating the config rather than going through the pipeline creation
Properties required for validation are right there in the config
Which fields need containsMacro() checks before validating is easier to see within the config
If another config extends that config, they can inherit the validation logic.

These things are only applicable if everything you need for validation resides in the config. If that's the case, I would move it to the config.

Reply:
I agree @rohitsinha54 @albertshau that we should have validate methods in config. However, in this particular case, we are verifying the compatibility of the output schema fields with bigquery columns. This means the validation is against bigquery fields and does not use many config variables. Thats why the logic was part of the plugin.

However, we can still move some logic to config as below.
``` java
 private void validateSchema(BigQuerySinkConfig config) throws IOException {
    Table table = getBigqueryTable(config);
    if (table == null) {
      // Table does not exist, so no further validation is required.
      return;
    }

    com.google.cloud.bigquery.Schema bqSchema = table.getDefinition().getSchema();
    if (bqSchema == null) {
      // Table is created without schema, so no further validation is required.
      return;
    }

    FieldList bqFields = bqSchema.getFields();
    List<Schema.Field> outputSchemaFields = config.getSchema().getFields();
    // Match compatibility of output schema fields with bigquery table fields
    config.matchSchemaCompatibility(outputSchemaFields, bqFields);
  }
``` 
and in config class
```java
/**
   * Matches the compatibility of schema fields with bigquery column types.
   *
   * @param schemaFields list of {@link Schema.Field}s to match compatibility with bigquery fields.
   * @param bqFields List of bigquery table fields
   * @throws IllegalArgumentException If output schema has fields that are not present in {@link FieldList} or
   * if one of the required fields is missing or type of {@link Schema.Field} does not match {@link Field} type.
   */
  public void matchSchemaCompatibility(List<Schema.Field> schemaFields, FieldList bqFields) {
}
```

what do you think?

Reply:
I don't think there's a need to split it up. Don't think this is a big deal let's keep it as is